README
                              FluidSynth-DSSI
                              ===============
                    A software synthesizer plugin for the
                         DSSI Soft Synth Interface

Introduction
============
FluidSynth-DSSI is an implementation of the FluidSynth soundfont-
playing software synthesizer as a plugin for the DSSI Soft Synth
Interface.  DSSI is a plugin API for software instruments (soft
synths) with user interfaces, permitting them to be hosted
in-process by audio applications.  More information on DSSI can be
found at:

  http://dssi.sourceforge.net/

FluidSynth-DSSI makes use of DSSI's run_multiple_synths() interface
to allow sharing of resources (like enormous soundfonts) between
multiple plugin instances -- soundfont data is shared between
instances, and FluidSynth's usual voice allocation methods are
applied across multiple instances as if each were a FluidSynth
channel.

FluidSynth-DSSI is written by Sean Bolton, and is copyright (c)2009
under the GNU General Public License, version 2 or later.  See the
enclosed file COPYING for details.

FluidSynth itself is copyright (c)2006 Peter Hanappe and others,
under the GNU Library General Public License.

Requirements
============
You need to have the following installed:

    - DSSI version 0.9 or greater, available from the
        dssi.sourceforge.net address above.

    - liblo version 0.12 or later (0.23 or later recommended), a
        library implementing the Open Sound Control (OSC) protocol,
        available at:

        http://liblo.sourceforge.net/

    - the FluidSynth library and header files, version 1.0.3 or later.
        FluidSynth-DSSI will have additional capability (run-time
        polyphony setting and more efficient operation) if FluidSynth
        version 1.0.6 or later is used.

    - the LADSPA v1.x SDK.

    - the ALSA headers (DSSI plugins use ALSA structures, but not
        the actual drivers, so you don't necessarily need the
        drivers installed.)  Users of non-Linux system can use
        libdssialsacompat, available at:

        http://smbolton.com/linux.html

    - pkgconfig with PKG_CONFIG_PATH set appropriately to pick up
        DSSI, liblo, and fluidsynth (plus libdssialsacompat if used).

    - glib-2.0 >= 2.40

    - GTK+ version 2.x.  (If GTK+ is not found, the plugin
        will be built without the GUI.)

    - a working DSSI host. Fluidsynth-DSSI has been tested with
        qtractor found at https://github.com/rncbc/qtractor

    - automake 1.7 and autoconf 2.59 or better if you wish to
        recreate the build files.

Installation
============
The generic installation instructions in the enclosed file INSTALL
aren't particularly helpful, so try this instead:

1. Unpack the tar file.

2. Make sure PKG_CONFIG_PATH is set correctly to locate the
    dssi.pc and liblo.pc pkgconfig files.  On many systems, this
    will be:

    $ PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
    $ export PKG_CONFIG_PATH

3. 'cd' into the package directory and execute './configure'.
    configure will add '-Wall' and my favorite optimizations to
    CFLAGS for you if you don't include any '-Wall' or '-O' options.

4. If you wish to change the maximum number of simultaneous
    instances supported by FluidSynth-DSSI (currently 32), edit the
    file src/fluidsynth-dssi.h and change the FSD_CHANNEL_COUNT
    definition.

3. Enable debugging information if you desire: edit the files
    src/fluidsynth-dssi.h and src/FluidSynth-DSSI_gtk.c, and define
    DEBUG_DSSI or DEBUG_AUDIO as explained in the comments.

4. Do 'make'.  Hopefully it should build without warnings (or
    errors.)

5. 'make install' will install the following:

    $(PREFIX)/lib/dssi/fluidsynth-dssi.so
    $(PREFIX)/lib/dssi/fluidsynth-dssi/FluidSynth-DSSI_gtk

Feedback on your experiences building FluidSynth-DSSI is appreciated.

Operation
=========
Start your DSSI host and create an instance of the FluidSynth-DSSI
plugin.

Initially, no soundfont will be loaded.  Use the 'Select Soundfont'
button to select a soundfont. Each instance may use only one
soundfont at a time, but when multiple instances use the same
soundfont, it is only loaded once.  FluidSynth-DSSI will use the
the DSSI project directory and the SF2_PATH enviroment variable,
if set, to search for soundfonts.

The available soundfont presets will be displayed in the 'Preset'
window.  Clicking on one will select that preset.

Under 'Global Settings', the 'gain (dB)' setting affects the output
level of all running FluidSynth-DSSI instances.

If you're using FluidSynth version 1.0.6 or later, then you will
have the ability to manage the maximum polyphony used by
FluidSynth-DSSI.  There will be a 'polyphony' setting under 'Global
Settings' in the FluidSynth-DSSI GUI, which sets the maximum number
of voices that will be played by all instances.

Test notes may be generated using the 'Send Test Note' button.  You
may select the pitch and velocity of the note using the 'Test Note'
sliders.

When you create more than one instance of the plugin, you should be
able to tell their GUIs apart by unique window titles of each
(provided your host correctly supplies each with the proper
identifier.)


ChangeLog
20180124 FluidSynth-DSSI v2.0.0 release

 * Add /usr/share/soundfonts to sf2 default paths
 * Fixes for recent autotools
 * Fixes for fluidsynth 2.x
 * Add configuration file support for
    * audio/realtime-prio: realtime priority for threads.
    * synth/maxchannels: Maximum channels supported by fluidsynth-dssi
    * parallel-threads: Number of parallel threads used for sound rendering
    * dynamic-sample-loading: -> synth.dynamic-sample-loading
 * Don't trash effects but set NULL to avoid useless copying within fluidsynth
    to reduce CPU load
 * Fix some gcc truncation warnings
 * Set default max channels to 32

20120324 CVS smbolton
 * Bug fix to fsd_cleanup() from Rui Nuno Capela, allowing eventual
     reinstantiation of FluidSynth (sf.net Tracker ID 1990241).
 * Patch from Albert Graef to pass sysex messages on to FluidSynth,
     allowing the use of MIDI Tuning Standard.
 * Fixed plugin to use reentrant strtok_r().

20090104 FluidSynth-DSSI v1.0.0 release
 * Version bump to 1.0.0.

20081214 CVS smbolton
 * Version bump to 0.9.1.2.
 * Bug fix from James McDermott for the broken-nwrite_float
     code in fluidsynth-dssi.c when used with unusual sample
     counts.
 * Removed the FluidSynth source patches in extra/, removed
     references to the patches in README, and reworked configure.ac
     to simply check that the FluidSynth version is >=1.0.6 before
     setting USE_AUGMENTED_FLUIDSYNTH_API and
     NWRITE_FLOAT_WORKS_CORRECTLY.
 * Added '-test' option to FluidSynth-DSSI_gtk.c to allow testing
     the GUI without a host.
 * Fixed to use gcc/glibc 'constructor' and 'destructor' attributes
    instead of the deprecated _init() and _fini() functions (needed
    for OS X 10.4 'Tiger').
 * Added code in FluidSynth-DSSI_gtk.c to silently ignore the
    'sample-rate' method.

20050608 CVS smbolton
 * Version bump to 0.9.1.1.
 * Updated code, patch, and docs to match Josh Green's FluidSynth
     CVS commit of 20050607.

20050530 FluidSynth-DSSI v0.9.1 release
 * Symlink fluidsynth-dssi to jack-dssi-host, if it exists.

20050510 CVS smbolton
 * Version bump to 0.9.1.
 * Moved FluidSynth-DSSI into its own SourceForge CVS module.
 * configure.ac improvements.

20050424 v0.9.1alpha2 smbolton
 * Changed to an API-conforming but less efficient implementation
     in the plugin using fluid_synth_nwrite_float().
 * Bundled the soundfont metadata loading code necessary to make
     the GUI independent of libfluidsynth.
 * Added automake/autoconf build system.

20041109
 * Implemented global polyphony setting.
 * GUI resizing fix.
 * Optionally use GTK+ v2.x (edit Makefile).

20041101
 * Updated for DSSI v0.9: implemented project directory.
 * Updated for liblo v0.12: removed non-blocking hack.
 * Implemented loading soundfont from SF2_PATH in GUI.
 * Implemented gain setting in GUI.
 * Better inter-thread cooperation: fsd_configure() now avoids
     grabbing the mutex on 'load' calls when the soundfont is
     already loaded, and fsd_select_program() checks for valid
     programs before grabbing mutex, and saves the program change
     if it can't get the lock.
 
20041005
 * GUI now waits to send OSC update request from inside gtk_main()

20040715 DSSI v0.4 release
 * Added minimal GTK+ GUI.
 * Require DSSI v0.4.
 * Incorporated into DSSI as example code.
 * GUI: raise on 'show' if already visible.
 * Makefile: added install, use of liblo pkgconfig, use of correct
     FluidSynth includes.

20040623 v0.0.2
 * Added support for per-instance soundfont selection.
 * Changed from rendering directly into the output buffers, to
     always rendering into 16-byte aligned temporary buffers then
     copying to the output, in case libfluidsynth.* was compiled
     with SSE support.
 * Added mutex protection of fluidsynth structures between audio and
     control threads.



src/locate_soundfont.c
/* FluidSynth DSSI software synthesizer plugin
 *
 * Copyright (C) 2004, 2009 Chris Cannam and others.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA.
 */

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <unistd.h>

#define DEFAULT_SF2PATH "/usr/local/share/sf2:/usr/share/sf2:/usr/share/soundfonts"

char *
fsd_locate_soundfont_file(const char *origpath, const char *projectDirectory)
{
    struct stat statbuf;
    char *sf2path, *path, *origPath, *element, *context, *eltpath;
    const char *filename;

    if (stat(origpath, &statbuf) == 0)
        return strdup(origpath);

    filename = strrchr(origpath, '/');

    if (filename) ++filename;
    else filename = origpath;
    if (!*filename) return NULL;

    sf2path = getenv("SF2_PATH");
    if (sf2path) path = strdup(sf2path);
    else {
	char *home = getenv("HOME");
	if (!home) path = strdup(DEFAULT_SF2PATH);
	else {
	    path = (char *)malloc(strlen(DEFAULT_SF2PATH) + strlen(home) + 6);
	    sprintf(path, "%s/sf2:%s", home, DEFAULT_SF2PATH);
	}
    }

    if (projectDirectory) {
	origPath = path;
	path = (char *)malloc(strlen(origPath) + strlen(projectDirectory) + 2);
	sprintf(path, "%s:%s", projectDirectory, origPath);
	free(origPath);
    }

    origPath = path;

    while ((element = strtok_r(path, ":", &context)) != 0) {

	path = 0;

	if (element[0] != '/') {
	    /* DEBUG_DSSI("fluidsynth-dssi: Ignoring relative element %s in path\n", element); */
	    continue;
	}

	eltpath = (char *)malloc(strlen(element) + strlen(filename) + 2);
	sprintf(eltpath, "%s/%s", element, filename);

	if (stat(eltpath, &statbuf) == 0) {
	    free(origPath);
	    return eltpath;
	}

	free(eltpath);
    }

    free(origPath);
    return NULL;
}

int
fsd_sf2_compare(const void *a, const void *b)
{
    const char *s1 = *(const char **)a;
    const char *s2 = *(const char **)b;
    return strcasecmp(s1, s2);
}    

char **
fsd_get_known_soundfonts(const char *projectDirectory, int *rn)
{
    int n = 0;
    int s = 0;
    char **list = 0;
    char *sf2path, *path, *origPath, *element;
    DIR *subdir;
    struct dirent *entry;
    int i;

    sf2path = getenv("SF2_PATH");
    if (sf2path) path = strdup(sf2path);
    else {
	char *home = getenv("HOME");
	if (!home) path = strdup(DEFAULT_SF2PATH);
	else {
	    path = (char *)malloc(strlen(DEFAULT_SF2PATH) + strlen(home) + 6);
	    sprintf(path, "%s/sf2:%s", home, DEFAULT_SF2PATH);
	}
    }

    if (projectDirectory) {
	origPath = path;
	path = (char *)malloc(strlen(origPath) + strlen(projectDirectory) + 2);
	sprintf(path, "%s:%s", projectDirectory, origPath);
	free(origPath);
    }

    origPath = path;

    while ((element = strtok(path, ":")) != 0) {

	path = 0;

	if (element[0] != '/') {
	    /* DEBUG_DSSI("fluidsynth-dssi: Ignoring relative element %s in path\n", element); */
	    continue;
	}

	if (!(subdir = opendir(element))) {
	    continue;
	}

	while ((entry = readdir(subdir))) {
	    
	    if (entry->d_name[0] == '.') continue;

	    if (strlen(entry->d_name) > 4 &&
		!strcasecmp(entry->d_name + strlen(entry->d_name) - 4, ".sf2")) {
		
		for (i = 0; i < n; ++i) {
		    if (!strcmp(entry->d_name, list[i])) break;
		}
		if (i == n) {
		    if (i == s) {
			s += 100;
			list = realloc(list, s * sizeof(char *));
		    }
		    list[n] = strdup(entry->d_name);
		    ++n;
		}
	    }
	}
    }

    *rn = n;
    free(origPath);

    if (n > 0) {
	qsort(list, n, sizeof(char *), fsd_sf2_compare);
    }

    return list;
}


src/load_soundfont_metadata.h
/* FluidSynth DSSI software synthesizer GUI
 *
 * Copyright (C) 2005, 2009 Sean Bolton and others.
 *
 * All of this code comes from FluidSynth, copyright (C) 2003 Peter
 * Hanappe and others, which in turn borrows code from GLIB,
 * copyright (C) 1995-1997 Peter Mattis, Spencer Kimball and Josh
 * MacDonald, and from Smurf SoundFont Editor, copyright (C) Josh
 * Green.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA.
 */

#ifndef _LOAD_SFONT_METADATA_H
#define _LOAD_SFONT_METADATA_H

#include <stdio.h>

/* ==== from fluid_list.h: ==== */

typedef struct _fluid_list_t fluid_list_t;

struct _fluid_list_t
{
  void* data;
  fluid_list_t *next;
};

#define fluid_list_next(slist)	((slist) ? (((fluid_list_t *)(slist))->next) : NULL)

/* ==== from fluid_defsfont.h: ==== */

/* Sound Font structure defines */

typedef struct _SFVersion
{                               /* version structure */
      unsigned short major;
      unsigned short minor;
}
SFVersion;

typedef struct _SFPreset
{                               /* Preset structure */
      char name[21];            /* preset name */
      unsigned short prenum;    /* preset number */
      unsigned short bank;      /* bank number */
      unsigned int libr;        /* Not used (preserved) */
      unsigned int genre;       /* Not used (preserved) */
      unsigned int morph;       /* Not used (preserved) */
      fluid_list_t *zone;       /* list of preset zones */
}
SFPreset;

typedef struct _SFData
{                               /* Sound font data structure */
  SFVersion version;            /* sound font version */
  SFVersion romver;             /* ROM version */
  unsigned int samplepos;       /* position within sffd of the sample chunk */
  unsigned int samplesize;      /* length within sffd of the sample chunk */
  char *fname;                  /* file name */
  FILE *sffd;                   /* loaded sfont file descriptor */
  fluid_list_t *info;           /* linked list of info strings (1st byte is ID) */
  fluid_list_t *preset;         /* linked list of preset info */
  fluid_list_t *inst;           /* linked list of instrument info */
  fluid_list_t *sample;         /* linked list of sample info */
}
SFData;

SFData *fsd_sfload_file (const char * fname);
void    fsd_sfont_free_data (SFData * sf);

#endif /* _LOAD_SFONT_METADATA_H */


src/fluidsynth-dssi.h
/* FluidSynth DSSI software synthesizer plugin
 *
 * Copyright (C) 2004-2005, 2009 Sean Bolton and others.
 *
 * Portions of this file may have come from Peter Hanappe's
 * Fluidsynth, copyright (C) 2003 Peter Hanappe and others.
 * Portions of this file may have come from Chris Cannam and Steve
 * Harris's public domain DSSI example code.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA.
 */

#ifndef _FLUIDSYNTH_DSSI_H
#define _FLUIDSYNTH_DSSI_H

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <ladspa.h>

#include <fluidsynth.h>

/* ==== debugging ==== */

/* DSSI interface debugging -- choose one: */
#define DEBUG_DSSI(fmt...)
// #define DEBUG_DSSI(fmt...) { fprintf(stderr, fmt); }

/* audio debugging -- define if desired: */
// #define DEBUG_AUDIO

/* ==== end of debugging ==== */

#define FSD_MAX_CHANNELS  255  /* FluidSynth's virtual channel limit (n.b. NO_CHANNEL in fluid_voice.h) */

/* FSD_CHANNEL_COUNT_DEFAULT is the number of channels FluidSynth-DSSI tells
 * libfluidsynth to create at startup.  It can be up to FSD_MAX_CHANNELS,
 * but be aware that for each channel, libfluidsynth will zero and copy out
 * two buffers every process cycle, so it's best to set FSD_CHANNEL_COUNT
 * to no more than the maximum number of simultaneous FluidSynth-DSSI
 * instances you'll need. */
#define FSD_CHANNEL_COUNT_DEFAULT  32

/* -FIX- These should be in a header file shared with FluidSynth-DSSI_gtk.c: */
#define FSD_MAX_POLYPHONY     256
#define FSD_DEFAULT_POLYPHONY 256

#define FSD_MAX_BURST_SIZE  128

typedef enum {
    PORT_OUTPUT_LEFT     = 0,
    PORT_OUTPUT_RIGHT    = 1,
    FSD_PORTS_COUNT      = 2
} PortNumbers;

struct fsd_port_descriptor {
    LADSPA_PortDescriptor          port_descriptor;
    char *                         name;
    LADSPA_PortRangeHintDescriptor hint_descriptor;
    LADSPA_Data                    lower_bound;
    LADSPA_Data                    upper_bound;
};

typedef struct _fsd_settings_t fsd_settings_t;
typedef struct _fsd_instance_t fsd_instance_t;
typedef struct _fsd_sfont_t fsd_sfont_t;
typedef struct _fsd_synth_t fsd_synth_t;

struct _fsd_settings_t {
    int channel_count;
    int thread_count;
    int realtime_prio;
#if FLUIDSYNTH_VERSION_MAJOR >= 2
    int dynamic_sample_loading;
#endif
};

struct _fsd_instance_t {
    int             channel;
    int             pending_preset_change;
    fsd_sfont_t    *soundfont;
    LADSPA_Data    *output_l;
    LADSPA_Data    *output_r;
};

struct _fsd_sfont_t {
    fsd_sfont_t             *next;
    char                    *path;
    int                      sfont_id;
    int                      ref_count;
    int                      preset_count;
    DSSI_Program_Descriptor *presets;
};

struct _fsd_synth_t {
    pthread_mutex_t   mutex;
    int               mutex_grab_failed;
    unsigned long     instance_count;
    char             *project_directory;
    fluid_settings_t *fluid_settings;
    fluid_synth_t    *fluid_synth;
    fsd_sfont_t      *soundfonts;
    float             gain;
    int               polyphony;
    fsd_instance_t   **channel_map;
    LADSPA_Data       bit_bucket[FSD_MAX_BURST_SIZE];
};

#endif /* _FLUIDSYNTH_DSSI_H */

